wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift

```diff
@@ -211,98 +211,238 @@ extension GenericMessage {
+        guard let encryptedData = encryptedData else {
+            return nil
+        }
 
-        return (data, missingClientsStrategy)
+        return (encryptedData, missingClientsStrategy)
     }
 
     /// Attempts to generate an encrypted payload for the given collection of user clients.
 
-    public func encryptForTransport(for recipients: [ZMUser: Set<UserClient>],
-                                    useQualifiedIdentifiers: Bool = false,
-                                    in context: NSManagedObjectContext) -> EncryptedPayloadGenerator.Payload? {
-
+    public func encryptForTransport(
+        for recipients: [ZMUser: Set<UserClient>],
+        useQualifiedIdentifiers: Bool = false,
+        in context: NSManagedObjectContext
+    ) -> EncryptedPayloadGenerator.Payload? {
         // We're targeting a specific client so we want to ignore all missing clients.
         let missingClientsStrategy = MissingClientsStrategy.ignoreAllMissingClients
+        let encryptedData: Data?
+
+        if DeveloperFlag.proteusViaCoreCrypto.isOn {
+            encryptedData = encrypt(
+                for: recipients,
+                with: missingClientsStrategy,
+                useQualifiedIdentifiers: useQualifiedIdentifiers,
+                in: context
+            )
+        } else {
+            encryptedData = legacyEncrypt(
+                for: recipients,
+                with: missingClientsStrategy,
+                useQualifiedIdentifiers: useQualifiedIdentifiers,
+                in: context
+            )
+        }
 
-        guard let data = encrypt(for: recipients,
-                                 with: missingClientsStrategy,
-                                 useQualifiedIdentifiers: useQualifiedIdentifiers,
-                                 in: context) else { return nil }
+        guard let encryptedData = encryptedData else {
+            return nil
+        }
 
-        return (data, missingClientsStrategy)
+        return (encryptedData, missingClientsStrategy)
     }
 
-    private func encrypt(for recipients: [ZMUser: Set<UserClient>],
-                         with missingClientsStrategy: MissingClientsStrategy,
-                         externalData: Data? = nil,
-                         useQualifiedIdentifiers: Bool = false,
-                         in context: NSManagedObjectContext) -> Data? {
+    private func encrypt(
+        for recipients: [ZMUser: Set<UserClient>],
+        with missingClientsStrategy: MissingClientsStrategy,
+        externalData: Data? = nil,
+        useQualifiedIdentifiers: Bool = false,
+        in context: NSManagedObjectContext
+    ) -> Data? {
+        guard
+            let proteusService = context.proteusService,
+            let selfClient = ZMUser.selfUser(in: context).selfClient(),
+            selfClient.remoteIdentifier != nil
+        else {
+            return nil
+        }
+
+        var messageData: Data?
+
+        // TODO: get core crypto file lock
```

Comment:
what does this TODO mean? Don't we have file locks already with safe core crypto?

Reply:
It's not in this branch yet.